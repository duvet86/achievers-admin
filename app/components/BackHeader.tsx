import { Link } from "@remix-run/react";

import { ArrowLeft } from "iconoir-react";

interface Props {
  to?: string;
}

export function BackHeader({ to = "../" }: Props) {
  return (
    <>
      <div>
        <Link to={to} relative="path" className="btn btn-ghost mb-2 gap-2">
          <ArrowLeft className="w-6" />
          Back
        </Link>
      </div>

      <hr className="mb-4" />
    </>
  );
}
