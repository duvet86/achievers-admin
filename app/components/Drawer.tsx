import { NavLink } from "@remix-run/react";

import { NavArrowRight, User, GraduationCap, Home } from "iconoir-react";

interface Props {
  isAdmin: boolean;
  version: string;
}

interface DrawerLink {
  icon: JSX.Element | null;
  value: string;
  label: string;
}

function getLinks(isAdmin: boolean): DrawerLink[] {
  return isAdmin
    ? [
        {
          icon: <User className="h-6 w-6" />,
          value: "/users",
          label: "Mentors",
        },
        {
          icon: <GraduationCap className="h-6 w-6" />,
          value: "/students",
          label: "Students",
        },
        {
          icon: <Home className="h-6 w-6" />,
          value: "/chapters",
          label: "Chapters",
        },
      ]
    : [
        {
          icon: null,
          value: "/roster",
          label: "Roster",
        },
        {
          icon: null,
          value: "/mentees",
          label: "My Mentees",
        },
      ];
}

export function Drawer({ isAdmin, version }: Props) {
  const links = getLinks(isAdmin);

  return (
    <div className="drawer-side pt-16">
      <label htmlFor="drawer" className="drawer-overlay"></label>
      <ul className="menu h-full w-80 border-r border-primary bg-base-200 p-4 text-base-content">
        {links.map(({ icon, label, value }, index) => (
          <li key={index}>
            <NavLink
              to={value}
              className={({ isActive }) =>
                isActive
                  ? "active mb-2 justify-between rounded font-semibold"
                  : "mb-2 justify-between rounded font-semibold"
              }
            >
              <div className="flex items-center gap-4">
                {icon}
                {label}
              </div>
              <NavArrowRight className="h-6 w-6" />
            </NavLink>
          </li>
        ))}
      </ul>
      <div
        data-testid="version"
        className="absolute bottom-0 left-2 z-10 text-sm italic"
      >
        Version {version}
      </div>
    </div>
  );
}
