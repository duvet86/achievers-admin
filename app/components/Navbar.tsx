import type { AzureUserWebAppWithRole, Environment } from "~/services";

import { Link } from "@remix-run/react";

import { ProfileCircle, Menu, LogOut } from "iconoir-react";

interface Props {
  currentUser: AzureUserWebAppWithRole;
  environment: Environment;
  version: string;
}

export function Navbar({ currentUser, environment, version }: Props) {
  const showEnvBadge = environment === "local" || environment === "staging";

  return (
    <nav className="navbar absolute left-0 top-0 z-10 h-16 bg-primary text-primary-content shadow-md">
      <div className="flex-none lg:hidden">
        <label htmlFor="drawer" className="btn btn-square btn-ghost">
          <Menu className="inline-block h-6 w-6 stroke-current" />
        </label>
      </div>
      <div className="flex-1">
        <Link to="/" className="btn btn-ghost text-xl normal-case">
          <img
            className="mr-4 w-16 rounded"
            src="/images/logo.png"
            alt="Logo"
          />
          <span className="text-xl font-semibold tracking-tight">
            Achievers WA
          </span>
          {showEnvBadge && (
            <span className="badge badge-lg ml-2">{environment}</span>
          )}
        </Link>
      </div>

      <div className="dropdown dropdown-end hidden lg:block">
        <div tabIndex={0} className="flex cursor-pointer items-center gap-2">
          <div className="font-semibold">{currentUser.email}</div>
          <label className="btn btn-circle btn-ghost">
            <div className="flex w-10 content-center justify-center rounded-full">
              <ProfileCircle />
            </div>
          </label>
        </div>
        <ul
          tabIndex={0}
          className="menu dropdown-content w-52 rounded-box bg-base-100 p-2 shadow"
        >
          <li>
            <Link className="font-semibold" to="/logout">
              <LogOut className="mr-2 w-6" />
              Logout
            </Link>
          </li>
          <li className="menu-title">
            <hr />
            <div className="mt-2 text-xs italic">Version {version}</div>
          </li>
        </ul>
      </div>
    </nav>
  );
}
