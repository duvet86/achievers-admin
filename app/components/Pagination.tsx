import classNames from "classnames";

import { FastArrowLeft, FastArrowRight, MoreHoriz } from "iconoir-react";

interface Props {
  range: number[];
  currentPageNumber: number;
  totalPageCount: number;
}

export function Pagination({
  range,
  currentPageNumber,
  totalPageCount,
}: Props) {
  const isPreviousDisabled = currentPageNumber === 0;
  const isNextDisabled =
    currentPageNumber === totalPageCount - 1 || totalPageCount === 0;

  return (
    <div className="join mt-4">
      <button
        type="submit"
        name="previousBtn"
        value="previousBtn"
        className={classNames("btn btn-outline join-item", {
          "btn-disabled": isPreviousDisabled,
        })}
        disabled={isPreviousDisabled}
        title="previous"
      >
        <FastArrowLeft />
      </button>
      {range.map((pageNumber, index) =>
        pageNumber === -1 ? (
          <button
            key={index}
            className="btn btn-disabled btn-outline join-item"
          >
            <MoreHoriz />
          </button>
        ) : (
          <button
            key={index}
            type="submit"
            name="pageNumberBtn"
            value={pageNumber - 1}
            className={classNames("btn btn-outline join-item", {
              "btn-active": currentPageNumber === pageNumber - 1,
            })}
          >
            {pageNumber}
          </button>
        ),
      )}
      <button
        type="submit"
        name="nextBtn"
        value="nextBtn"
        className={classNames("btn btn-outline join-item", {
          "btn-disabled": isNextDisabled,
        })}
        disabled={isNextDisabled}
        title="next"
      >
        <FastArrowRight />
      </button>
    </div>
  );
}
