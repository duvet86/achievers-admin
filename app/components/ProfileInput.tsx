import { useRef, useState } from "react";

import { Cancel } from "iconoir-react";

interface Props {
  defaultValue: string | null;
}

export default function ProfileInput({ defaultValue }: Props) {
  const inputRef = useRef<HTMLInputElement | null>(null);

  const [profilePicturePath, setProfilePicturePath] = useState<string | null>(
    defaultValue,
  );
  const [selectedImage, setSelectedImage] = useState<File | null>(null);

  const hasImage = selectedImage || profilePicturePath;

  return (
    <div className="card card-side m-8 bg-base-100 shadow-xl">
      <figure>
        <img
          src={
            selectedImage
              ? URL.createObjectURL(selectedImage)
              : profilePicturePath
              ? profilePicturePath
              : "/images/profile-picture.webp"
          }
          alt="profile"
          className="h-44 w-44"
        />
      </figure>
      <div className="card-body">
        <h2 className="card-title mb-4">Profile picture</h2>

        {hasImage && (
          <button
            className="btn btn-error w-32 gap-2"
            onClick={(e) => {
              e.preventDefault();
              if (inputRef.current) {
                inputRef.current.files = null;
                inputRef.current.value = "";
              }
              setSelectedImage(null);
              setProfilePicturePath(null);
            }}
          >
            <Cancel className="h-6 w-6" />
            Remove
          </button>
        )}

        <input
          type="file"
          name="profilePicture"
          className={hasImage ? "hidden" : "file-input w-full max-w-xs"}
          ref={inputRef}
          accept="image/png, image/gif, image/jpeg"
          onChange={(event) => {
            setSelectedImage(event.target.files?.[0] ?? null);
          }}
        />

        <input
          type="hidden"
          name="deleteProfilePicture"
          value={(!hasImage).toString()}
        />
      </div>
    </div>
  );
}
