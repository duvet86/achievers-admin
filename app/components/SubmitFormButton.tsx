import { useEffect, useState } from "react";
import classnames from "classnames";

import { Check, FloppyDiskArrowIn, WarningCircle } from "iconoir-react";

interface Props {
  label?: string;
  successMessage: string | null | undefined;
  errorMessage?: string | null;
  sticky?: boolean;
  className?: string;
}

export function SubmitFormButton({
  successMessage,
  errorMessage,
  sticky,
  className,
  label = "Save",
}: Props) {
  return (
    <div
      data-testid="container"
      className={classnames("flex", className, {
        "sticky bottom-0": sticky,
      })}
    >
      <Message
        key={Date.now()}
        errorMessage={errorMessage}
        successMessage={successMessage}
      />

      <button className="btn btn-primary w-52 gap-5" type="submit">
        <FloppyDiskArrowIn className="h-6 w-6" />
        {label}
      </button>
    </div>
  );
}

interface MessageProps {
  successMessage: string | undefined | null;
  errorMessage: string | undefined | null;
}

function Message({ successMessage, errorMessage }: MessageProps) {
  const [isActive, setIsActive] = useState(true);

  useEffect(() => {
    let n: number | undefined;
    if (isActive) {
      n = window.setTimeout(() => {
        setIsActive(false);
      }, 3000);
    }

    return () => {
      if (n) {
        clearTimeout(n);
      }
    };
  }, [isActive]);

  return (
    <div data-testid="message">
      {isActive && successMessage && (
        <div className="alert alert-success pr-12">
          <Check />
          <span>{successMessage}</span>
        </div>
      )}
      {errorMessage && (
        <div className="alert alert-error pr-12">
          <WarningCircle />
          <span>{errorMessage}</span>
        </div>
      )}
    </div>
  );
}
