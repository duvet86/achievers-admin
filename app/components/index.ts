export * from "./checkbox/Checkbox";
export * from "./date-input/DateInput";
export * from "./input/Input";
export * from "./textarea/Textarea";
export * from "./select/Select";
export * from "./file-input/FileInput";
export * from "./submit-form-button/SubmitFormButton";
export * from "./radio/Radio";
export * from "./autocomplete/Autocomplete";

export * from "./title/Title";
export * from "./sub-title/SubTitle";
export * from "./text/Text";
export * from "./back-header/BackHeader";
export * from "./pagination/Pagination";

export * from "./loading-spinner/LoadingSpinner";
export * from "./navbar/Navbar";
export * from "./drawer/Drawer";
export * from "./body/Body";

export * from "./not-found/NotFound";
export * from "./forbidden/Forbidden";
export * from "./unexpected-error/UnexpectedError";

export * from "./editor/Editor";
