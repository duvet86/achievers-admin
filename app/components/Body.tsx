import type { AzureUserWebAppWithRole, Environment } from "~/services";

import { Outlet } from "@remix-run/react";

import { Drawer } from "./Drawer";
import { Navbar } from "./Navbar";

interface Props {
  isAdmin: boolean;
  hasCompletedVolunteerAgreement?: boolean;
  version: string;
  currentUser: AzureUserWebAppWithRole;
  environment: Environment;
}

export function Body({
  currentUser,
  hasCompletedVolunteerAgreement,
  isAdmin,
  version,
  environment,
}: Props) {
  const showDrawer = isAdmin || hasCompletedVolunteerAgreement;

  return (
    <div className="drawer lg:drawer-open">
      <input id="drawer" type="checkbox" className="drawer-toggle" />
      <div className="drawer-content flex flex-col">
        <Navbar currentUser={currentUser} environment={environment} />

        <main className="mt-16 flex h-full flex-col overflow-y-auto bg-white p-4">
          <Outlet />
        </main>
      </div>

      {showDrawer && <Drawer isAdmin={isAdmin} version={version} />}
    </div>
  );
}
