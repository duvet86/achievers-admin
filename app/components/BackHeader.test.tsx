import "@testing-library/jest-dom";

import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router";

import { BackHeader } from "~/components";

describe("BackHeader", () => {
  it("BackHeader snapshot", async () => {
    const { baseElement } = render(
      <MemoryRouter>
        <BackHeader to="/" />
      </MemoryRouter>,
    );

    expect(baseElement).toMatchSnapshot();
  });
});
