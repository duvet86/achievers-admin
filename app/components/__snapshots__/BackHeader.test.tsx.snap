// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackHeader > BackHeader snapshot 1`] = `
<body>
  <div>
    <div>
      <button
        class="btn btn-ghost mb-2 gap-2"
      >
        <svg
          class="w-6"
          color="currentColor"
          fill="none"
          height="1.5em"
          stroke-width="1.5"
          viewBox="0 0 24 24"
          width="1.5em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M21 12L3 12M3 12L11.5 3.5M3 12L11.5 20.5"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
          />
        </svg>
        Back
      </button>
    </div>
    <hr
      class="mb-4"
    />
  </div>
</body>
`;
